README.md
# Format Microformat

[![Build Status](https://travis-ci.org/voxpelli/node-format-microformat.svg?branch=master)](https://travis-ci.org/voxpelli/node-format-microformat)
[![Coverage Status](https://coveralls.io/repos/voxpelli/node-format-microformat/badge.svg)](https://coveralls.io/r/voxpelli/node-format-microformat)
[![dependencies Status](https://david-dm.org/voxpelli/node-format-microformat/status.svg)](https://david-dm.org/voxpelli/node-format-microformat)
[![Known Vulnerabilities](https://snyk.io/test/github/voxpelli/node-format-microformat/badge.svg?targetFile=package.json)](https://snyk.io/test/github/voxpelli/node-format-microformat?targetFile=package.json)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat)](https://github.com/Flet/semistandard)

Formats a Microformat JSON representation into eg. a Jekyll post

## Requirements

Requires at least Node.js 6.x

## Installation

```bash
npm install format-microformat --save
```

## Current status

**Alpha**

Currently formats data into a hardcoded Jekyll style. This is intended to become more dynamic and configurable in the future to adapt to more styles.

## Usage

Simple:

```javascript
const MicropubFormatter = require('format-microformat');

const formatter = new MicropubFormatter();

formatter.formatAll(micropubDocument)
  .then(formatted => {
    // Lots of formatted data that one can do lots of fun stuff with. Publish somewhere or such perhaps?
    // Available keys are "filename", "url", "content" and lastly "files" if any files were uploaded
  });
```

Advanced:

```javascript
const MicropubFormatter = require('format-microformat');

const formatter = new MicropubFormatter('http://example.com/');

formatter.preFormat(micropubDocument)
  .then(preFormatted => Promise.all([
    formatter.formatFilename(preFormatted),
    formatter.format(preFormatted),
    formatter.formatURL(preFormatted),
  ]))
  .then(([formattedFilename, formattedContent, formattedUrl]) => {
    // Lots of formatted data that you can do lots of fun stuff with. Publish it somewhere maybe?
  });
```

## Constructor

`new MicropubFormatter([options])`

### Options

* **relativeTo** – if set to a URL, then all formatted URL:s will be resolved to absolute URL:s relative to that one rather then be returned as relative ones.
* **noMarkdown** – if set to `true` then no conversion to Markdown will happen for the content.
* **contentSlug** – if set to `true`, then the slug creation will use the `properties.content` data as a fallback to `properties.name` prior to basing the slug on the timestamp.
* **defaults** – a `micropubDocument` with defaults that will be added as part of the `preFormat()`. Useful to eg. ensure that all documents have a language explicitly set.
* **deriveCategory** – a method that's provided the properties of the post and which returns a dervice category – or `false` to disable default category deriving
* **deriveLanguages** – an array defining what languages, using [ISO 639-3](https://en.wikipedia.org/wiki/ISO_639-3), to autodetect – or `true` to try and autodetect everything
* **filenameStyle** – a filename style in the same form as the Jekyll [permalink style](https://jekyllrb.com/docs/permalinks/). Should not include file extension. Defaults to: `_posts/:year-:month-:day-:slug`
* **filesStyle** – the filename style of uploaded files, in an extended form of the Jekyll [permalink style](https://jekyllrb.com/docs/permalinks/). Should not include file extension, but should include `:filesslug`. Defaults to: `media/:year-:month-:slug/:filesslug`
* **layoutName** – the name of a layout that overrides the default `micropubpost` one, or `false` to remove the layout completely from the front matter
* **permalinkStyle** – a Jekyll [permalink style](https://jekyllrb.com/docs/permalinks/). Defaults to Jekyll's default: `date`
* **encodeHTML** – _non-standard_ – if set to `false` then HTML-encoding will not happen for the content value. Defaults to `true`

`layoutName`, `filenameStyle`, `filesStyle` and `permalinkStyle` can all be set directly or through a callback that's given some data and that callback might either return a value directly or return a `Promise` that eventually resolves to the value.

## Methods

* **formatAll(micropubDocument)** – `preFormat`:s and formats everything. Returns a `Promise`that resolves to an object with the keys `filename`, `url`, `content`, `files` and `raw`.
* **preFormat(micropubDocument)** – takes a `micropubDocument` and ensures that all necessary parts are there. **Currently required** to run a `micropubDocument` through this method before handing it to the rest of the methods (except `formatAll()`).
* **formatFilename(preformattedMicropubDocument)**  – returns a filename based on the data in the `micropubDocument`. Includes the relative path to the file – which currently is always `_posts/`
* **formatURL(preformattedMicropubDocument)**  – returns the url the formatted content is expected to live on when published
* **format(preformattedMicropubDocument)** – formats the actual content. Currently it's formatted as an HTML-file with [Jekyll Front Matter](http://jekyllrb.com/docs/frontmatter/). The content of the file is the `properties.content` of the `micropubDocument` – the rest of the data is put into the *front matter*.

## Output formats

For exact implementation of how things are formatted, look at the code and the tests, this is just to get an overview of what one can expect the output to be.

### Filename

The target is to get a filename similar to `_posts/2015-06-30-awesomeness-is-awesome.html`. The date first and then either the defined slug or a slug derived from title or content.

### URL

The target is to get a relatuve URL similar to `2015/06/awesomeness-is-awesome/` where the slug is calculated in the same way as it is for the `filename`. In some cases the URL will be prefixed with a category name – that's the case for eg. replies, likes and bookmarks – the first two are prefixed with `interaction/` and the last one with `bookmark/`.

### Content

The actual content of the file is the HTML of the `properties.content` of the `micropubDocument`. All other properties are added to the [Jekyll Front Matter](http://jekyllrb.com/docs/frontmatter/) together with a couple of other defaults.

Some `micropubDocument` properties receive special handling, the rest are included raw as arrays with an `mf-` prefix to avoid collisions with pre-existing Jekull properties.

The ones with special handling are:

* **content** – isn't part of the front matter but is the actual HTML content of the file
* **name** – inserted as `title` and if not specified, then a value of `''` will be used to avoid automatic generation of titles
* **slug** – inserted as `slug`
* **category** – inserted as `tags`
* **published** – inserted as `date` and formatted as ISO-format

There's also some defaults and derived values:

* **layout** – by default set to `micropubpost`
* **category** – derived from other `micropubDocument` properties and only used in some cases, like eg. for replies, likes and bookmarks – it's set to  `interaction`  for the first two and to `bookmark` for the last one

An example of a generated Jekyll Front Matter:

```yaml
layout: micropubpost
date: '2015-06-30T14:34:01.000Z'
title: awesomeness is awesome
slug: awesomeness-is-awesome
category: interaction
mf-like-of:
  - 'http://example.com/liked/page'
```

### Files

An array of objects with a `filename` containing the full path where the file should be uploaded as well as a `buffer` key containing the same buffer object that was part of the original `micropubDocument` files data. Also calculates URL:s for the location of the files post-upload and adds them to the proper `photo`, `video` or `audio` property in the `micropubDocument` so that they are made available in the `content`.

Unlike the other formatting, file formatting happens fully in `preFormat()` as it needs to be done prior to the rest of the step to make the URL:s of the uploaded files available to the rest of the formatting.

Currently supported file keys from the original `micropubDocument`: `photo`, `video` and `audio`

## Format of `micropubDocument`

The format closely matches the [JSON-representation](http://indiewebcamp.com/Micropub#JSON_Syntax) of Micropub.

See the [micropub-express](https://github.com/voxpelli/node-micropub-express#format-of-micropubdocument) module for documentation of this basic object.

In addition to the properties defined by the `micropub-express` module, the `preFormat()` method adds a top level `derived` key that's used internally for values derived from.

The `preFormat()` also flattens the `files` object into an array of files and formats the filenames to the full path where a file should be uploaded.

## Other useful modules

* [micropub-express](https://github.com/voxpelli/node-micropub-express) – an Express 4 Micropub endpoint that accepts and verifies Micropub requests and calls a callback with a parsed `micropubDocument` that can be used with this module
* [github-publish](https://github.com/voxpelli/node-github-publish) – a module that takes a filename and content and publishes that to a GitHub repository. A useful place to send the formatted data that comes out of this module if one wants to add it to a GitHub hosted Jekyll blog of some kind, like eg. [GitHub Pages](https://pages.github.com/).

## Used in

* [webpage-micropub-to-github](https://github.com/voxpelli/webpage-micropub-to-github) – a self-hosteable Micropub endpoint that publishes posts to Jekyll sites by committing them to their GitHub repositories


CHANGELOG.md
## 0.11.1 (2019-01-20)

* **Feature:** Non-standard: Made it possible to opt out of HTML encoding of plain text content through new `encodeHTML` option

## 0.11.0 (2018-12-25)

* **Breaking change:** Now requires Node v8

## 0.10.2 (2018-01-20)

* **Feature:** Made post layout configurable through `layoutName`
* **Bug fix:** Fixed issue with sending the same object more than once to `preFormat()`

## 0.10.1 (2017-05-16)

* **Improvement:** Make `filenameStyle`, `filesStyle` and `permalinkStyle` configurable through a callback

## 0.10.0 (2017-05-14)

* **Feature:** Made post's filename configurable through `filenameStyle`
* **Feature:** Made uploaded files' filename style configurable through `filesStyle`

## 0.9.1 (2017-01-21)

* **Improvement:** By default publish all posts in the past by 15 seconds to avoid time sync issues

## 0.9.0 (2016-10-23)

* **Breaking change:** Now requires Node v6
* **Improvements:** Updated dev dependencies and moved to a Grunt-less, [semistandard](https://github.com/Flet/semistandard)-based setup through [ESLint](http://eslint.org/)
* **Improvements:** Updated Travis definition and test targets
* **Minor:** Added `yarn.lock` to `.gitignore` as this is a library and [libraries don't use lock files](https://github.com/yarnpkg/yarn/issues/838#issuecomment-253362537)

## 0.8.0 (2016-08-17)


#### Features

* **main:** enable custom derived category ([15e832b4](https://github.com/voxpelli/node-format-microformat/commit/15e832b4dfa90886d3c8a08707b9cac65268bbd8))


## 0.7.0 (2016-07-16)


#### Features

* **main:** made permalink style configurable ([2c987995](https://github.com/voxpelli/node-format-microformat/commit/2c98799531cb860b71423f583c8518a38a9f97e7))


#### Breaking Changes

* now uses Jekyll's default permalink style instead of "/:categories/:year/:month/:title/" as previously
 ([2c987995](https://github.com/voxpelli/node-format-microformat/commit/2c98799531cb860b71423f583c8518a38a9f97e7))
* New minimum Node.js version: 5.x
 ([1e1c50b4](https://github.com/voxpelli/node-format-microformat/commit/1e1c50b45156645a9888497a9b9d3a521dc7b8c2))


## 0.6.0 (2016-03-09)


#### Features

* **main:** Jekyll 3 compatibility ([8040c422](https://github.com/voxpelli/node-format-microformat/commit/8040c422c355cf576bad6b5b361ad86dc3b2cf1f))


### 0.5.1 (2015-09-07)


#### Bug Fixes

* **main:** some bugs with new type of HTML-content ([9d131e04](https://github.com/voxpelli/node-format-microformat/commit/9d131e04de3e2d25bc6cc078f6a9fc63e976068b))


## 0.5.0 (2015-08-28)


#### Features

* **main:** require content.html for html content ([962e66a8](https://github.com/voxpelli/node-format-microformat/commit/962e66a89533f614a5b6f2583d8480157847da29))


### 0.4.1 (2015-08-07)


#### Bug Fixes

* **main:** should detect undetectable language ([b5134b5f](https://github.com/voxpelli/node-format-microformat/commit/b5134b5fbc1d5f782e6aee6c2823effe64b2f6a9))


#### Features

* **main:** extract person tags from categories ([c245871f](https://github.com/voxpelli/node-format-microformat/commit/c245871f2e18fe514436cfbdad1ef240bea615c9))


## 0.4.0 (2015-07-26)


#### Bug Fixes

* **main:** rename "link" to "links" ([bd30567c](https://github.com/voxpelli/node-format-microformat/commit/bd30567c57e0caeadd1ee9dc848cd2f08842b9ce))


#### Features

* **main:** optin: if no language, autodetect it ([bbfc567b](https://github.com/voxpelli/node-format-microformat/commit/bbfc567bf52154240a511ea3c94d8bc76dcb6b43))


### 0.3.2 (2015-07-24)


#### Bug Fixes

* **main:**
  * preformat publish time before slug ([b8c7813b](https://github.com/voxpelli/node-format-microformat/commit/b8c7813b14bf0848f40d456afd5b1c0564ff8f11))
  * by default don't base slug on content ([affb0f73](https://github.com/voxpelli/node-format-microformat/commit/affb0f7361e32feac9098ada20c7f4cf310eb8f7))
  * avoid dashes in beginning/end of slug ([39508870](https://github.com/voxpelli/node-format-microformat/commit/39508870ed929c9eb11eb4404ab0ca381c39b653))


#### Features

* **main:**
  * expose language in top Front Matter ([1de72fb7](https://github.com/voxpelli/node-format-microformat/commit/1de72fb7c95809b2f3ed77853d0de32653940249))
  * make it possible to define defaults ([072455ef](https://github.com/voxpelli/node-format-microformat/commit/072455ef7c6776ba9502bb56be294b771f50107b))


### 0.3.1 (2015-07-22)


#### Bug Fixes

* **main:**
  * categorize "notes" as social activity ([5a991355](https://github.com/voxpelli/node-format-microformat/commit/5a9913557251df85e0e98a2b477126ee7ae5de22))
  * ignore url properties ([4c8b007b](https://github.com/voxpelli/node-format-microformat/commit/4c8b007b253eb1baeb23e5a6b5659f8cafe4b933))


#### Features

* **main:** expose preformatted data in formatAll ([f0fd23f6](https://github.com/voxpelli/node-format-microformat/commit/f0fd23f62220eb5a41c3b17f08a37e5cfbe28c9b))


## 0.3.0 (2015-07-21)


#### Features

* **main:** convert HTML to markdown ([721b3523](https://github.com/voxpelli/node-format-microformat/commit/721b352377f09509c99ad84b0ab42c94ce705b49))


### 0.2.2 (2015-07-21)


#### Bug Fixes

* **main:**
  * map interactions to other categories ([6d617888](https://github.com/voxpelli/node-format-microformat/commit/6d617888b25a738d42504e4c733efc81fe044911))
  * slug creation now handles foreign chars ([87f3d80a](https://github.com/voxpelli/node-format-microformat/commit/87f3d80a9765b1c8ec272b76d82edfc061529077))
  * improved kebab casing ([874a1af7](https://github.com/voxpelli/node-format-microformat/commit/874a1af70e6bb2bd5f94aa10470aaae0a1176a40))
  * support the "bookmark-of" as well ([d793dc89](https://github.com/voxpelli/node-format-microformat/commit/d793dc89c9f8596bb17f93d7467005f3ce26f209))


### 0.2.1 (2015-07-21)


#### Bug Fixes

* **main:** restore original full date in filenames ([073512fe](https://github.com/voxpelli/node-format-microformat/commit/073512fe41cf9cc9557a58df23e777a5f2789b08))


## 0.2.0 (2015-07-20)


#### Bug Fixes

* **main:**

test/fixtures.js
'use strict';

module.exports = function () {
  const baseMicroformatData = {
    'type': ['h-entry'],
    'properties': {
      'content': ['hello world'],
      'name': ['awesomeness is awesome'],
      'slug': ['awesomeness-is-awesome'],
      'published': [new Date(1435674841000)]
    }
  };

  return {
    baseMicroformatData
  };
};


test/url.spec.js
'use strict';

const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');
const sinon = require('sinon');
const sinonChai = require('sinon-chai');

chai.use(chaiAsPromised);
chai.use(sinonChai);
chai.should();

describe('URL', () => {
  const Formatter = require('../');
  const getFixtures = require('./fixtures');

  let formatter;
  let baseMicroformatData;

  beforeEach(() => {
    const fixtures = getFixtures();

    formatter = new Formatter({
      permalinkStyle: '/:categories/:year/:month/:title/'
    });
    baseMicroformatData = fixtures.baseMicroformatData;
  });

  afterEach(() => {
    sinon.restore();
  });

  describe('formatURL', () => {
    it('should base URL on slug', () => {
      return formatter.formatURL(baseMicroformatData).should.eventually.equal('2015/06/awesomeness-is-awesome/');
    });

    it('should support undefined permalink style', () => {
      formatter = new Formatter();
      return formatter.formatURL(baseMicroformatData).should.eventually.equal('2015/06/30/awesomeness-is-awesome.html');
    });

    it('should return absolute URL when requested', () => {
      formatter = new Formatter('http://example.com/foo/');
      return formatter.formatURL(baseMicroformatData).should.eventually.equal('http://example.com/foo/2015/06/30/awesomeness-is-awesome.html');
    });

    it('should include derived category if any', () => {
      baseMicroformatData.derived = { category: 'interaction' };
      return formatter.formatURL(baseMicroformatData).should.eventually.equal('interaction/2015/06/awesomeness-is-awesome/');
    });

    it('should support name in permalink style', () => {
      formatter = new Formatter({
        permalinkStyle: '/:name'
      });
      return formatter.formatURL(baseMicroformatData).should.eventually.equal('2015-06-30-awesomeness-is-awesome');
    });

    it('should support custom permalink style through callback', () => {
      const permalinkStyle = sinon.stub().returns('/:name');

      formatter = new Formatter({ permalinkStyle });

      return formatter.formatURL(baseMicroformatData).should.eventually.equal('2015-06-30-awesomeness-is-awesome');
    });

    it('should support custom permalink style through callback returning Promise', () => {
      const permalinkStyle = sinon.stub().resolves('/:name');

      formatter = new Formatter({ permalinkStyle });

      return formatter.formatURL(baseMicroformatData).should.eventually.equal('2015-06-30-awesomeness-is-awesome');
    });
  });
});


test/slug.spec.js
'use strict';

const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');

chai.use(chaiAsPromised);
chai.should();

describe('Slug', () => {
  const Formatter = require('../');
  const getFixtures = require('./fixtures');

  let formatter;
  let baseMicroformatData;

  beforeEach(() => {
    const fixtures = getFixtures();

    formatter = new Formatter();
    baseMicroformatData = fixtures.baseMicroformatData;
  });

  describe('_formatSlug', () => {
    beforeEach(() => {
      delete baseMicroformatData.properties.slug;
    });

    it('should base slug on title', () => {
      formatter._formatSlug(baseMicroformatData).should.equal('awesomeness-is-awesome');
    });

    it('should fall back to base slug on content', () => {
      delete baseMicroformatData.properties.name;
      formatter = new Formatter({ contentSlug: true });
      formatter._formatSlug(baseMicroformatData).should.equal('hello-world');
    });

    it('should ignore html-tags when basing slug on content', () => {
      delete baseMicroformatData.properties.name;
      baseMicroformatData.properties.content = [{ html: '<h1>Foo</h1> Bar &amp; <strong>Abc</strong>' }];
      formatter = new Formatter({ contentSlug: true });
      // Test twice so that we don't get a non-reusable regexp!
      formatter._formatSlug(baseMicroformatData).should.equal('foo-bar-abc');
      formatter._formatSlug(baseMicroformatData).should.equal('foo-bar-abc');
    });

    it('should ultimately fall back to publish time', () => {
      delete baseMicroformatData.properties.name;
      formatter._formatSlug(baseMicroformatData).should.equal('52441');
    });

    it('should limit length of extracted slug', () => {
      baseMicroformatData.properties.name = ['One Two Three Four Five Six Seven'];
      formatter._formatSlug(baseMicroformatData).should.equal('one-two-three-four-five');

      baseMicroformatData.properties.content = baseMicroformatData.properties.name;
      delete baseMicroformatData.properties.name;
      formatter = new Formatter({ contentSlug: true });
      formatter._formatSlug(baseMicroformatData).should.equal('one-two-three-four-five');
    });

    it('should properly format abbreviations', () => {
      baseMicroformatData.properties.name = ['Another CSS-feature is the FooBar'];
      formatter._formatSlug(baseMicroformatData).should.equal('another-css-feature-is-the-foo-bar');
    });

    it('should properly handle swedish chars', () => {
      baseMicroformatData.properties.name = ['ÖverÄnda på Slottet'];
      formatter._formatSlug(baseMicroformatData).should.equal('over-anda-pa-slottet');
    });

    it('should ensure slug doesnt start or end with a dash', () => {
      delete baseMicroformatData.properties.name;
      baseMicroformatData.properties.content = [',One Two Three Four Five, Six Seven'];
      formatter = new Formatter({ contentSlug: true });
      formatter._formatSlug(baseMicroformatData).should.equal('one-two-three-four-five');
    });
  });
});


index.js
'use strict';

const pathModule = require('path');
const { URL } = require('url');
const escapeHtml = require('escape-html');
const yaml = require('js-yaml');
const ent = require('ent');
const Upndown = require('upndown');
const franc = require('franc');
const iso6393 = require('iso-639-3');
const jekyllUtils = require('jekyll-utils');

const deburr = require('lodash.deburr');
const defaultsDeep = require('lodash.defaultsdeep');
const cloneDeepWith = require('lodash.clonedeepwith');

const htmlRegexp = /<[^>]+>/g;
const camelRegexp = /([a-z])([A-Z])/g;
const kebabRegexp = /[^a-z0-9]+/g;
const whitespaceRegexp = /\s+/g;
const httpRegexp = /^http(s?):\/\//;

const isEmpty = function (value) {
  if (!value) { return true; }
  if (Array.isArray(value) && value.length === 0) { return true; }
  if (typeof value === 'object' && Object.keys(value).length === 0) { return true; }
  return false;
};

const getMfValue = function (data) {
  return (data || [])
    .map(item => {
      if (!item) { return; }
      if (item.value) { return item.value; }
      if (item.html) {
        return ent.decode(item.html.replace(htmlRegexp, ' '))
          .replace(whitespaceRegexp, ' ')
          .trim();
      }
      if (typeof item === 'string') { return item; }
    })
    .filter(item => !!item);
};

const semiKebabCase = function (name) {
  // Convert camel case to spaces, then ensure everything is lower case and then finally – make kebab
  return deburr(name)
    .replace(camelRegexp, '$1 $2')
    .trim()
    .toLowerCase()
    .replace(kebabRegexp, ' ')
    .trim()
    .replace(whitespaceRegexp, '-');
};

const zipObject = function (keys, values) {
  const result = {};

  keys.forEach((key, i) => {
    result[key] = values[i];
  });

  return result;
};

const objectPromiseAll = function (obj, mapMethod) {
  const keys = Object.keys(obj);
  return Promise.all(keys.map(key => mapMethod ? mapMethod(obj[key], key) : obj[key]))
    .then(result => zipObject(keys, result));
};

const resolveValue = function (value, arg) {
  if (typeof value === 'function') {
    value = value(arg);
  }

  return Promise.resolve(value);
};

const resolveCallbackParameter = function (value, data) {
  return resolveValue(value, (data.formattedData || data).properties);
};

const Formatter = function (options) {
  if (typeof options === 'string') {
    options = { relativeTo: options };
  } else {
    options = Object.assign({}, options || {});
  }

  this.relativeTo = options.relativeTo;
  this.markdown = !options.noMarkdown;
  this.contentSlug = !!options.contentSlug;
  this.defaults = options.defaults;
  this.deriveLanguages = options.deriveLanguages || false;
  this.filenameStyle = options.filenameStyle || '_posts/:year-:month-:day-:slug';
  this.filesStyle = options.filesStyle || 'media/:year-:month-:slug/:filesslug';
  this.permalinkStyle = options.permalinkStyle;
  this.deriveCategory = options.deriveCategory === undefined ? true : options.deriveCategory;
  this.layoutName = options.layoutName;
  this.encodeHTML = options.encodeHTML === undefined ? true : options.encodeHTML;

  if (typeof this.filenameStyle !== 'function' && !this.filenameStyle.includes(':')) {
    throw new Error('Invalid filenameStyle, must include a placeholder');
  }

  if (typeof this.filesStyle !== 'function' && !this.filesStyle.includes(':filesslug')) {
    throw new Error('Invalid filesStyle, must include :filesslug');
  }
};

Formatter.prototype._resolveFrontMatterData = function (data) {
  const source = data.properties;
  const derived = data.derived || {};

  // TODO: Include the "type" property so that more than the now assumed "h-entry" can be supported

  const target = {
    layout: derived.layout || 'micropubpost',
    date: source.published[0].toISOString(),
    title: ''
  };

  const mapping = {
    name: 'title',
    slug: 'slug',
    category: 'tags',
    lang: 'lang'
  };
  const ignore = ['content', 'published', 'url'];

  Object.keys(source).forEach(key => {
    const value = source[key];

    if (!value.length || ignore.indexOf(key) !== -1) { return; }

    if (mapping[key]) {
      target[mapping[key]] = value.join(' ');
    } else {
      target['mf-' + key] = value;
    }
  });

  if (derived.layout === false) {
    delete target.layout;
  }
  if (derived.category) {
    target.category = derived.category;
  }
  if (derived.personTags) {
    target.persontags = derived.personTags;
  }

  return target;
};

Formatter.prototype._formatFrontMatter = function (data) {
  const frontMatterData = data.frontMatterData || this._resolveFrontMatterData(data);
  return '---\n' + yaml.safeDump(frontMatterData) + '---\n';
};

Formatter.prototype._formatContent = function (data) {
  data = data.formattedData || data;

  let content = data.properties.content;

  if (!content) { return Promise.resolve(''); }

  content = [].concat(content);

  const und = new Upndown();

  if (Array.isArray(content)) {
    return Promise.all(content.map(content => {
      if (typeof content !== 'object') {
        content = { value: content };
      }

      if (content.html) {
        return this.markdown ? new Promise(resolve => {
          und.convert(content.html, (err, markdown) => {
            resolve(err ? content.html : markdown);
          });
        }) : content.html;
      }

      return this.encodeHTML ? escapeHtml(content.value || '') : content.value || '';
    }))
      .then(result => result.filter(value => !!value).join('\n') + '\n');
  }
};

Formatter.prototype._formatSlug = function (data) {
  if (data.properties.slug && data.properties.slug[0]) {
    return semiKebabCase(data.properties.slug[0]);
  }

  let name;

  if (data.properties.name) {
    name = data.properties.name[0].trim();
  }
  if (!name && !isEmpty(data.properties.content) && this.contentSlug) {
    name = getMfValue(data.properties.content).join('\n');
  }

  if (name) {
    name = name.split(/\s+/);
    if (name.length > 5) {
      name = name.slice(0, 5);
    }
    name = name.join(' ');
  } else {
    name = Math.floor(data.properties.published[0].getTime() / 1000) % (24 * 60 * 60) + '';
  }

  return semiKebabCase(name);
};

Formatter.prototype._preFormatFiles = function (data) {
  const fileResolves = [];

  ['video', 'photo', 'audio'].forEach(type => {
    (data.files[type] || []).forEach(file => {
      fileResolves.push(Promise.all([
        type,
        file,
        this.formatFilesFilename(type, file, data),
        this.formatFilesURL(type, file, data)
      ]));
    });
  });

  return Promise.all(fileResolves).then(files => {
    const newFiles = [];

    files.forEach(file => {
      const type = file[0];
      const fileData = file[1];
      const filename = file[2];
      const url = file[3];

      data.properties[type] = data.properties[type] || [];
      data.properties[type].push(url);

      newFiles.push({
        filename: filename,
        buffer: fileData.buffer
      });
    });

    data.files = newFiles;

    return data;
  });
};

Formatter.prototype.preFormat = function (data) {
  if (data.preFormatted) {
    return Promise.resolve(data);
  }

  data = cloneDeepWith(data, value => value instanceof Buffer ? value : undefined);

  data.preFormatted = true;

  data.properties.published = [
    data.properties.published && data.properties.published[0]
      ? new Date(data.properties.published[0])
      // Default to a time 15 seconds into the past to avoid issues with time being out of sync between
      // the Micropub server and the build server, that has proven to be an issue otherwise.
      : new Date(Date.now() - 15000)
  ];

  const slug = this._formatSlug(data);
  data.properties.slug = slug ? [slug] : [];

  let strippedContent, estimatedLang;
  if (isEmpty(data.properties.lang) && this.deriveLanguages && !isEmpty(data.properties.content)) {
    strippedContent = getMfValue(data.properties.content).join('\n');

    if (strippedContent !== '') {
      estimatedLang = franc(strippedContent, this.deriveLanguages === true ? {} : { whitelist: this.deriveLanguages });
    }

    if (estimatedLang && estimatedLang !== 'und' && (this.deriveLanguages === true || this.deriveLanguages.indexOf(estimatedLang) !== -1)) {
      data.properties.lang = [estimatedLang];
    }
  }

  if (Array.isArray(data.properties.lang)) {
    data.properties.lang = data.properties.lang.map(lang => {
      if (!lang || lang.length !== 3) { return lang; }

      const code = iso6393.get(lang);

      if (code) {
        return code.iso6391;
      }

      return lang;
    });
  }

  data.derived = {};

  if (!isEmpty(data.properties.category)) {
    data.derived.personTags = [];

    data.properties.category = data.properties.category.filter(tag => {
      if (httpRegexp.test(tag)) {
        data.derived.personTags.push(tag);
        return false;
      }
      return true;
    });

    if (isEmpty(data.derived.personTags)) {
      delete data.derived.personTags;
    }
  }

  if (typeof this.deriveCategory === 'function') {
    data.derived.category = this.deriveCategory(data.properties);
    if (!data.derived.category) { delete data.derived.category; }
  } else if (!this.deriveCategory) {
    // Do nothing
  } else if (
    !isEmpty(data.properties.bookmark) ||
    !isEmpty(data.properties['repost-of']) ||
    !isEmpty(data.properties['bookmark-of'])
  ) {
    data.derived.category = 'links';
  } else if (
    isEmpty(data.properties.name) || // This means it's not an "article", but a "note"
    !isEmpty(data.properties['in-reply-to']) ||
    !isEmpty(data.properties['like-of'])
  ) {
    data.derived.category = 'social';
  }

  defaultsDeep(data, this.defaults || {});

  let result = Promise.resolve(data);

  if (this.layoutName !== undefined) {
    result = result.then(data => {
      return resolveCallbackParameter(this.layoutName, data)
        .then(layout => {
          if (layout === false) {
            data.derived.layout = false;
          } else if (layout) {
            data.derived.layout = layout + '';
          }
          return data;
        });
    });
  }

  if (!isEmpty(data.files)) {
    result = result.then(this._preFormatFiles.bind(this));
  }

  return result;
};

Formatter.prototype._precalculate = function (formattedData, { skipFilename } = {}) {
  return Promise.all([
    skipFilename ? undefined : this.formatFilename(formattedData),
    this._resolveFrontMatterData(formattedData)
  ])
    .then(result => ({
      formattedData,
      filename: result[0],
      frontMatterData: result[1]
    }));
};

Formatter.prototype.format = function (data) {
  return Promise.all([
    this._formatFrontMatter(data),
    this._formatContent(data)
  ])
    .then(result => result.join(''));
};

Formatter.prototype._getFileExtension = function () {
  return this.markdown ? 'md' : 'html';
};

Formatter.prototype.formatFilename = function (data) {
  return objectPromiseAll({
    jekyllResource: this._getJekyllResource(data, { skipFilename: true }),
    filenameStyle: resolveCallbackParameter(this.filenameStyle, data)
  })
    .then(({ jekyllResource, filenameStyle }) =>
      jekyllUtils.generateUrl(filenameStyle, jekyllResource).replace(/^\/+/, '') +
      '.' +
      this._getFileExtension()
    );
};

Formatter.prototype._getJekyllResource = function (data, { skipFilename } = {}) {
  return Promise.resolve(data.formattedData ? data : this._precalculate(data, { skipFilename }))
    .then(data => ({
      basename_without_ext: data.filename ? pathModule.basename(data.filename, this._getFileExtension()) : data.frontMatterData.slug,
      date: data.formattedData.properties.published[0],
      data: {
        categories: data.frontMatterData.category ? [].concat(data.frontMatterData.category) : [],
        slug: data.frontMatterData.slug
      }
    }));
};

Formatter.prototype.formatURL = function (data) {
  return objectPromiseAll({
    jekyllResource: this._getJekyllResource(data),
    permalinkStyle: resolveCallbackParameter(this.permalinkStyle, data)
  })
    .then(({ jekyllResource, permalinkStyle }) => {
      let url = jekyllUtils.generateUrl(permalinkStyle, jekyllResource).replace(/^\/+/, '');

      if (this.relativeTo) {
        url = (new URL(url, this.relativeTo)).toString();
      }

      return url;
    });
};

Formatter.prototype._formatFilesSlug = function (type, file) {
  return file.filename.trim().split('.').map(name => semiKebabCase(name)).join('.');
};

Formatter.prototype.formatFilesFilename = function (type, file, data) {
  return objectPromiseAll({
    jekyllResource: this._getJekyllResource(data, { skipFilename: true }),
    filesStyle: resolveCallbackParameter(this.filesStyle, data)
  })
    .then(({ jekyllResource, filesStyle }) => ({
      jekyllResource,
      filesStyle: filesStyle.split(':filesslug').join(this._formatFilesSlug(type, file))
    }))
    .then(({ jekyllResource, filesStyle }) => jekyllUtils.generateUrl(filesStyle, jekyllResource).replace(/^\/+/, ''));
};

Formatter.prototype.formatFilesURL = function (type, file, data) {
  return this.formatFilesFilename(type, file, data)
    .then(url => this.relativeTo ? (new URL(url, this.relativeTo)).toString() : url);
};

Formatter.prototype.formatAll = function (data) {
  return this.preFormat(data)
    .then(formattedData => this._precalculate(formattedData))
    .then(result => Promise.all([
      this.formatFilename(result),
      this.formatURL(result, this.relativeTo),
      this.format(result),
      result.formattedData.files,
      result.formattedData
    ]))
    .then(result => ({
      filename: result[0],
      url: result[1],
      content: result[2],
      files: result[3],
      raw: result[4]
    }));
};

module.exports = Formatter;
